File: apps/web/src/lib/api-client.ts
import type {
  PaginatedResponse,
  FeedResponse,
  ArticleResponse,
  BookmarkResponse,
  NotificationResponse,
  SettingsResponse,
  CreateFeedRequest,
  CreateBookmarkRequest,
  ListArticlesQuery,
  SearchBookmarksQuery,
  UpdateSettingsRequest,
  PaginationParams,
  ImportOpmlResponse,
  ErrorResponse,
} from '~/types/api'
import { API_BASE_URL, NOTIFICATION_BASE_URL } from './constants'
import { fetchJwt } from './auth'

class ApiError extends Error {
  constructor(
    public readonly code: string,
    message: string,
    public readonly status: number,
  ) {
    super(message)
    this.name = 'ApiError'
  }
}

function resolveBaseUrl(path: string): string {
  if (path.startsWith('/notifications')) {
    return NOTIFICATION_BASE_URL
  }
  return API_BASE_URL
}

async function request<T>(
  path: string,
  options: RequestInit = {},
  retry = true,
): Promise<T> {
  const baseUrl = resolveBaseUrl(path)
  const token = await fetchJwt()

  const res = await fetch(`${baseUrl}${path}`, {
    ...options,
    headers: {
      ...options.headers,
      Authorization: `Bearer ${token}`,
    },
  })

  if (res.status === 401 && retry) {
    const newToken = await fetchJwt(true)
    const retryRes = await fetch(`${baseUrl}${path}`, {
      ...options,
      headers: {
        ...options.headers,
        Authorization: `Bearer ${newToken}`,
      },
    })

    if (!retryRes.ok) {
      const err: ErrorResponse = await retryRes.json().catch(() => ({
        error: { code: 'UNKNOWN', message: retryRes.statusText },
      }))
      throw new ApiError(err.error.code, err.error.message, retryRes.status)
    }

    if (retryRes.status === 204) return undefined as T
    return retryRes.json() as Promise<T>
  }

  if (!res.ok) {
    const err: ErrorResponse = await res.json().catch(() => ({
      error: { code: 'UNKNOWN', message: res.statusText },
    }))
    throw new ApiError(err.error.code, err.error.message, res.status)
  }

  if (res.status === 204) return undefined as T
  return res.json() as Promise<T>
}

function toQueryString(params: Record<string, unknown>): string {
  const entries = Object.entries(params).filter(
    ([, v]) => v !== undefined && v !== null && v !== '',
  )
  if (entries.length === 0) return ''
  const searchParams = new URLSearchParams()
  for (const [key, value] of entries) {
    searchParams.set(key, String(value))
  }
  return `?${searchParams.toString()}`
}

export const apiClient = {
  getFeeds(): Promise<readonly FeedResponse[]> {
    return request<FeedResponse[]>('/feeds')
  },

  createFeed(body: CreateFeedRequest): Promise<FeedResponse> {
    return request<FeedResponse>('/feeds', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    })
  },

  deleteFeed(id: string): Promise<void> {
    return request<void>(`/feeds/${encodeURIComponent(id)}`, { method: 'DELETE' })
  },

  importOpml(file: File): Promise<ImportOpmlResponse> {
    const formData = new FormData()
    formData.append('file', file)
    return request<ImportOpmlResponse>('/feeds/import-opml', {
      method: 'POST',
      body: formData,
    })
  },

  getArticles(
    params: ListArticlesQuery = {},
  ): Promise<PaginatedResponse<ArticleResponse>> {
    return request<PaginatedResponse<ArticleResponse>>(
      `/articles${toQueryString(params)}`,
    )
  },

  getArticle(id: string): Promise<ArticleResponse> {
    return request<ArticleResponse>(`/articles/${encodeURIComponent(id)}`)
  },

  getBookmarks(
    params: PaginationParams = {},
  ): Promise<PaginatedResponse<BookmarkResponse>> {
    return request<PaginatedResponse<BookmarkResponse>>(
      `/bookmarks${toQueryString(params)}`,
    )
  },

  getBookmark(id: string): Promise<BookmarkResponse> {
    return request<BookmarkResponse>(`/bookmarks/${encodeURIComponent(id)}`)
  },

  createBookmark(body: CreateBookmarkRequest): Promise<BookmarkResponse> {
    return request<BookmarkResponse>('/bookmarks', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    })
  },

  deleteBookmark(id: string): Promise<void> {
    return request<void>(`/bookmarks/${encodeURIComponent(id)}`, { method: 'DELETE' })
  },

  searchBookmarks(
    params: SearchBookmarksQuery,
  ): Promise<PaginatedResponse<BookmarkResponse>> {
    return request<PaginatedResponse<BookmarkResponse>>(
      `/bookmarks/search${toQueryString(params)}`,
    )
  },

  getNotifications(
    params: PaginationParams = {},
  ): Promise<PaginatedResponse<NotificationResponse>> {
    return request<PaginatedResponse<NotificationResponse>>(
      `/notifications${toQueryString(params)}`,
    )
  },

  markNotificationRead(id: string): Promise<NotificationResponse> {
    return request<NotificationResponse>(`/notifications/${encodeURIComponent(id)}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ is_read: true }),
    })
  },

  getSettings(): Promise<SettingsResponse> {
    return request<SettingsResponse>('/settings')
  },

  updateSettings(body: UpdateSettingsRequest): Promise<SettingsResponse> {
    return request<SettingsResponse>('/settings', {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    })
  },
} as const

File: apps/cli/src/client/api_client.rs
use std::path::Path;
use std::time::Duration;

use anyhow::{Context, Result};
use reqwest::Client;
use serde::de::DeserializeOwned;
use serde::Serialize;

use crate::client::models::{
    ArticleResponse, BookmarkResponse, CreateBookmarkRequest, CreateFeedRequest, ErrorResponseBody,
    FeedResponse, ImportOpmlResponse, PaginatedResponse,
};
use crate::error::ApiError;

pub struct ApiClient {
    client: Client,
    api_url: String,
    token: String,
}

impl ApiClient {
    pub fn new(api_url: &str, token: &str) -> Result<Self> {
        let client = Client::builder()
            .timeout(Duration::from_secs(30))
            .build()
            .context("Failed to build HTTP client")?;

        Ok(Self {
            client,
            api_url: api_url.to_string(),
            token: token.to_string(),
        })
    }

    async fn handle_error_response(&self, response: reqwest::Response) -> anyhow::Error {
        let status = response.status().as_u16();

        if let Ok(body) = response.json::<ErrorResponseBody>().await {
            ApiError {
                code: body.error.code,
                message: body.error.message,
                status,
            }
            .into()
        } else {
            ApiError {
                code: "UNKNOWN".to_string(),
                message: format!("Request failed with status {status}"),
                status,
            }
            .into()
        }
    }

    async fn get<T: DeserializeOwned>(&self, path: &str) -> Result<T> {
        let url = format!("{}{}", self.api_url, path);
        let response = self
            .client
            .get(&url)
            .bearer_auth(&self.token)
            .send()
            .await
            .context("Failed to send request")?;

        if !response.status().is_success() {
            return Err(self.handle_error_response(response).await);
        }

        response
            .json::<T>()
            .await
            .context("Failed to parse response")
    }

    async fn post<B: Serialize, T: DeserializeOwned>(&self, path: &str, body: &B) -> Result<T> {
        let url = format!("{}{}", self.api_url, path);
        let response = self
            .client
            .post(&url)
            .bearer_auth(&self.token)
            .json(body)
            .send()
            .await
            .context("Failed to send request")?;

        if !response.status().is_success() {
            return Err(self.handle_error_response(response).await);
        }

        response
            .json::<T>()
            .await
            .context("Failed to parse response")
    }

    async fn delete(&self, path: &str) -> Result<()> {
        let url = format!("{}{}", self.api_url, path);
        let response = self
            .client
            .delete(&url)
            .bearer_auth(&self.token)
            .send()
            .await
            .context("Failed to send request")?;

        if !response.status().is_success() {
            return Err(self.handle_error_response(response).await);
        }

        Ok(())
    }

    async fn post_multipart<T: DeserializeOwned>(
        &self,
        path: &str,
        form: reqwest::multipart::Form,
    ) -> Result<T> {
        let url = format!("{}{}", self.api_url, path);
        let response = self
            .client
            .post(&url)
            .bearer_auth(&self.token)
            .multipart(form)
            .send()
            .await
            .context("Failed to send request")?;

        if !response.status().is_success() {
            return Err(self.handle_error_response(response).await);
        }

        response
            .json::<T>()
            .await
            .context("Failed to parse response")
    }

    // --- Feed API ---

    pub async fn list_feeds(&self) -> Result<Vec<FeedResponse>> {
        self.get("/feeds").await
    }

    pub async fn create_feed(&self, url: &str) -> Result<FeedResponse> {
        let body = CreateFeedRequest {
            url: url.to_string(),
        };
        self.post("/feeds", &body).await
    }

    pub async fn delete_feed(&self, id: &str) -> Result<()> {
        self.delete(&format!("/feeds/{id}")).await
    }

    pub async fn import_opml(&self, file_path: &Path) -> Result<ImportOpmlResponse> {
        let file_bytes = tokio::fs::read(file_path)
            .await
            .context("Failed to read OPML file")?;

        let file_name = file_path
            .file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_else(|| "import.opml".to_string());

        let part = reqwest::multipart::Part::bytes(file_bytes)
            .file_name(file_name)
            .mime_str("application/xml")
            .context("Failed to set MIME type")?;

        let form = reqwest::multipart::Form::new().part("file", part);

        self.post_multipart("/feeds/import-opml", form).await
    }

    // --- Article API ---

    pub async fn list_articles(
        &self,
        feed_id: Option<&str>,
        unread: bool,
        page: u32,
        limit: u32,
    ) -> Result<PaginatedResponse<ArticleResponse>> {
        let mut params = vec![format!("page={page}"), format!("limit={limit}")];

        if let Some(fid) = feed_id {
            params.push(format!("feed_id={fid}"));
        }

        if unread {
            params.push("is_read=false".to_string());
        }

        let query = params.join("&");
        self.get(&format!("/articles?{query}")).await
    }

    pub async fn get_article(&self, id: &str) -> Result<ArticleResponse> {
        self.get(&format!("/articles/{id}")).await
    }

    // --- Bookmark API ---

    pub async fn list_bookmarks(
        &self,
        page: u32,
        limit: u32,
    ) -> Result<PaginatedResponse<BookmarkResponse>> {
        self.get(&format!("/bookmarks?page={page}&limit={limit}"))
            .await
    }

    pub async fn create_bookmark(
        &self,
        article_id: Option<&str>,
        url: Option<&str>,
    ) -> Result<BookmarkResponse> {
        let body = CreateBookmarkRequest {
            article_id: article_id.map(String::from),
            url: url.map(String::from),
        };
        self.post("/bookmarks", &body).await
    }

    pub async fn delete_bookmark(&self, id: &str) -> Result<()> {
        self.delete(&format!("/bookmarks/{id}")).await
    }

    pub async fn get_bookmark(&self, id: &str) -> Result<BookmarkResponse> {
        self.get(&format!("/bookmarks/{id}")).await
    }

    pub async fn search_bookmarks(
        &self,
        query: &str,
        page: u32,
        limit: u32,
    ) -> Result<PaginatedResponse<BookmarkResponse>> {
        let encoded_query = urlencoding::encode(query);
        self.get(&format!(
            "/bookmarks/search?q={encoded_query}&page={page}&limit={limit}"
        ))
        .await
    }
}
